network/local.go

```diff
@@ -277,27 +277,47 @@ func (lc *LocalConn) Type() ConnType {
 // All operations are thread-safe.
 // The messages are marshalled and stored in the queue as a slice of bytes.
 type connQueue struct {
-	*sync.Cond
-	queue  [][]byte
-	closed bool
+	incomingQueue chan []byte
+	outgoingQueue chan []byte
+	closeCh       chan bool
+	wg            sync.WaitGroup
 }
 
+// LocalMaxBuffer is the number of packets that can be sent simultaneously to the
+// same address.
+const LocalMaxBuffer = 200
+
 func newConnQueue() *connQueue {
-	return &connQueue{
-		Cond: sync.NewCond(&sync.Mutex{}),
+	cq := &connQueue{
+		incomingQueue: make(chan []byte, LocalMaxBuffer),
+		outgoingQueue: make(chan []byte, LocalMaxBuffer),
+		closeCh:       make(chan bool),
+	}
+	cq.wg.Add(1)
+	go cq.start()
+	return cq
+}
+
+func (c *connQueue) start() {
+	defer c.wg.Done()
+	for {
+		select {
+		case buff := <-c.incomingQueue:
+			c.outgoingQueue <- buff
+		case <-c.closeCh:
+			// to signal that the conn is closed
+			close(c.outgoingQueue)
+			close(c.incomingQueue)
+			return
+		}
 	}
 }
 
 // push inserts a packet in the queue.
-// push won't work if the connQueue is already closed and silently return.
+// push should not be called if the connQueue is closed. Since it's the
+// LocalManager's reponsability, it is enforced.
 func (c *connQueue) push(buff []byte) {
-	c.L.Lock()
-	defer c.L.Unlock()
-	if c.closed {
-		return
-	}
-	c.queue = append(c.queue, buff)
-	c.Signal()
+	c.incomingQueue <- buff
```

Comment:
I think we need to include _some_ cleanup (very objective measurement) whenever we can. Because we already have tons of __to be done later__ issues. Perhaps not here, but at least for me some continous refactoring is always good. In some measure, of course. But here you can perhaps open an issue.